wallet: Pass in transactions and messages for signing instead of exporting the private keys

achow101 opened the pull request:
Following #17261, the way to sign transactions, PSBTs, and messages was to use `GetSigningProvider()` and get a `SigningProvider` containing the private keys. However this may not be feasible for future `ScriptPubKeyMan`s, such as for hardware wallets. Instead of exporting a `SigningProvider` containing private keys, we need to pass these things into the `ScriptPubKeyMan` (via `CWallet`) so that they can do whatever is needed internally to sign them. This is largely a refactor as the logic of processing transactions, PSBTs, and messages for is moved into `LegacyScriptPubKeyMan` and `CWallet` instead of being handled by the caller (e.g. `signrawtransaction`).

To help with this, I've refactored the 3(!) implementations of a `SignTransaction()` function into one generic one. This function will be called by `signrawtransactionwithkey` and `LegacyScriptPubKeyMan::SignTransaction()`. `CWallet::CreateTransaction()` is changed to call `CWallet::SignTransaction()` which in turn, calls `LegacyScriptPubKeyMan::SignTransaction()`. Other `ScriptPubKeyMan`s may implement `SignTransaction()` differently.

`FillPSBT()` is moved to be a member function of `CWallet` and the `psbtwallet.cpp/h` files removed. It is further split so that `CWallet` handles filling the UTXOs while the `ScriptPubKeyMan` handles adding keys, derivation paths, scripts, and signatures. In the end `LegacyScriptPubKeyMan::FillPSBT` still calls `SignPSBTInput`, but the `SigningProvider` is internal to `LegacyScriptPubKeyMan`. Other `ScriptPubKeyMan`s may do something different.

A new `SignMessage()` function is added to both `CWallet` and `ScriptPubKeyMan`. Instead of having the caller (i.e. `signmessage` or the sign message dialog) get the private key, hash the message, and sign, `ScriptPubKeyMan` will now handle that (`CWallet` passes through to the `ScriptPubKeyMan`s as it does for many functions). This signing code is thus consolidated into `LegacyScriptPubKeyMan::SignMessage()`, though other `ScriptPubKeyMan`s may implement it differently. Additionally, a `SigningError` enum is introduced for the different errors that we expect to see from `SignMessage()`.

Lastly, `GetSigningProvider()` is renamed to `GetPublicSigningProvider()`. It will now only provide pubkeys, key origins, and scripts. `LegacySigningProvider` has it's `GetKey` and `HaveKey` functions changed to only return false. Future implementations should return `HidingSigningProvider`s where private keys are hidden.

Other things like `dumpprivkey` and `dumpwallet` are not changed because they directly need and access the `LegacyScriptPubKeyMan` so are not relevant to future changes.

achow101 committed:
Refactor rawtransaction's SignTransaction into generic SignTransaction function

achow101 committed:
Add SignTransaction function to ScriptPubKeyMan and LegacyScriptPubKeyMan

achow101 committed:
Implement CWallet::SignTransaction using ScriptPubKeyMan::SignTransaction

achow101 committed:
Use CWallet::SignTransaction in CreateTransaction and signrawtransactionwithwallet

Instead of duplicating signing code, just use the function we already
have.

achow101 committed:
Move FillPSBT to be a member of CWallet

achow101 committed:
Move key and script filling and signing from CWallet::FillPSBT to ScriptPubKeyMan::FillPSBT

Instead of fetching a SigningProvider from ScriptPubKeyMan in order
to fill and sign the keys and scripts for a PSBT, just pass that
PSBT to a new FillPSBT function that does all that for us.

achow101 committed:
Replace GetSigningProvider with GetSolvingProvider

Not all ScriptPubKeyMans will be able to provide private keys,
but pubkeys and scripts should be. So only provide public-only
SigningProviders, i.e. ones that can help with Solving.

achow101 committed:
Clear any input_errors for an input after it is signed

Make sure that there are no errors set for an input after it is signed.
This is useful for when there are multiple ScriptPubKeyMans. Some may
fail to sign, but one may be able to sign, and after it does, we don't
want there to be any more errors there.

Sjors commented:
@kallewoof may find this interesting too in light of BIP-322 (Generic signed message format) #16440

vasild commented on src/qt/signverifymessagedialog.cpp:
```diff
@@ -133,30 +133,31 @@ void SignVerifyMessageDialog::on_signMessageButton_SM_clicked()
+        case SigningError::OK:
+            error = tr("No error");
+        case SigningError::PRIVATE_KEY_NOT_AVAILABLE:
+            error = tr("Private key for the entered address is not available.");
```
Missing `brake;`?

Sjors replied:
Also these strings can be obtained from`SigningErrorString` in `error.cpp` (the wording here seems better)

achow101 replied:
The strings from `SigningErrorString` won't be translated (we don't translate things for RPC), so that logic had to be duplicated here. Also, the strings are different to maintain compatibility.

achow101 replied:
Added the breaks in.

vasild commented:
https://github.com/achow101/bitcoin/blob/e4714e0/src/rpc/misc.cpp#L340 still calls the old signing code `key.SignCompact();` directly. Maybe that needs to be updated to also use the newly added function `SignMessage()`?

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2403,34 +2403,32 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+{
+    // Sign the tx with ScriptPubKeyMans
+    bool result = false;
+    for (const auto& spk_man_pair : m_spk_managers) {
```
In  e496e2981699ac2938b74297d9b0bce355d26ad8 `Implement CWallet::SignTransaction using ScriptPubKeyMan::SignTransaction` at `// Sign the tx with ScriptPubKeyMans`, check `CanProvide()` first?

achow101 replied:
I've updated it to use `GetScriptPubKeyMan` which calls `CanProvide`

This change is slightly more complex but it does reduce the number of repeated signings.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2403,34 +2403,32 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+    // Sign the tx with ScriptPubKeyMans
+    bool result = false;
+    for (const auto& spk_man_pair : m_spk_managers) {
+        result |= spk_man_pair.second->SignTransaction(tx, coins, sighash, input_errors);
```
The way `input_errors` are set, if there are multiple matching `ScriptPubKeyMans`* they would override each-others results. The last commit e4714e05ca9f3000384f9ed8f44e91a485f88bee clears the error for any given input, but only if SPKman SignTransaction() returns true, which requires all input errors to be empty, so that seems useless. At minimum this is confusing.

`*` = not the case with legacy, and in practice shouldn't happen with descriptor wallets, but that's not enforced

achow101 replied:
The expectation is that all of the errors will be the same across all failed `ScriptPubKeyMan`s because they don't have the private keys. The ScriptPubKeyMan` which fully signs an input will clear the error for that input. This clearing is not dependent on `input_errors.empty()`.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2460,44 +2459,20 @@ TransactionError CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bool& comp
-            continue;
+    // Fill in information from ScriptPubKeyMans
+    for (const auto& spk_man_pair : m_spk_managers) {
+        TransactionError res = spk_man_pair.second->FillPSBT(psbtx, sighash_type, sign, bip32derivs);
```
In a90a98c3aa90ed4dc4f995fac0859e58f177a79a under `// Fill in information from ScriptPubKeyMans`, add `CanProvide()`?

achow101 replied:
I've updated it to use `GetScriptPubKeyMan` which calls `CanProvide`

This change is slightly more complex but it does reduce the number of repeated signings.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -503,6 +503,52 @@ bool LegacyScriptPubKeyMan::SignTransaction(CMutableTransaction& tx, const std::
+            return TransactionError::SIGHASH_MISMATCH;
+        }
+
+        // Get the scriptPubKey to know which SigningProvider to use
```
In a90a98c3aa90ed4dc4f995fac0859e58f177a79a: this comment is outdated, because you're no longer calling `GetSigningProvider(script, sigdata)`. Also the `CScript script` below is not used.

achow101 replied:
Fixed the comment and simplified this block to just check for correct `non_witness_utxo` or `continue` if there is no utxo.

vasild commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -203,13 +205,20 @@ class ScriptPubKeyMan
+    /** Creates new signatures and adds them to the transaction. Returns whether all inputs were signed */
+    virtual bool SignTransaction(CMutableTransaction& tx, const std::map<COutPoint, Coin>& coins, int sighash, std::map<int, std::string>& input_errors) const { return false; }
+    /** Sign a message with the given script */
+    virtual SigningError SignMessage(const std::string& message, const PKHash& pkhash, std::string& str_sig) const { return SigningError::OK; };
```
It would be better to return some error from here (like the neighboring methods do). In case somebody forgets to override this method, it would be unexpected to return OK and leave the output `str_sig` unmodified.

Maybe even use a pure virtual method `...str_sig) const = 0;` to cause a compilation error if somebody forgets to override `SignMessage()`. I checked that:
* it compiles if changed to pure virtual and
* it does not compile if changed to pure virtual and the `LegacyScriptPubKeyMan` child class forgets to override `SignMessage()`.

achow101 replied:
Changed it to return `SigningError::SIGNING_FAILED`. I'm avoiding pure virtual functions in `ScriptPubKeyMan` because there is no requirement for every `ScriptPubKeyMan` to implement every function.

Sjors reviewed (commented):
Concept ACK. I roughly reviewed code for e4714e05ca9f3000384f9ed8f44e91a485f88bee and left some comments.

0eb285f7146ae9beec3a19829cc4420ad9f251ae `Refactor rawtransaction's SignTransaction into generic SignTransaction function` would be easier to review (`--color-moved=dimmed-zebra -w`) if you renamed the variables in separate commit.

It may be better to combine 4595326628bd4c91c753e923535178fca9ade2fd with be7991f63de2c0cce7d1e622a6ee12dd8f7204c9 so the signing code is moved, rather than just appear (though it's short enough to compare manually).

> Lastly, `GetSigningProvider()` is renamed to `GetPublicSigningProvider()`. It will now only provide pubkeys, key origins, and scripts. `LegacySigningProvider` has it's `GetKey` and `HaveKey` functions changed to only return false. Future implementations should return `HidingSigningProvider`s where private keys are hidden.

This bit is fairly confusing.

This PR is fairly large, and only deduplicates a subset of @vasild's #17577 (only what's needed to move forward with the wallet refactor). I didn't study this other PR in detail, but if it's possible to build on top of that, it would split the review burden and remove more duplication.

achow101 commented on src/script/sign.cpp:
```diff
@@ -465,3 +465,51 @@ bool IsSegWitOutput(const SigningProvider& provider, const CScript& script)
+
+        // amount must be specified for valid segwit signature
+        if (amount == MAX_MONEY && !txin.scriptWitness.IsNull()) {
+            input_errors[i] = "Missing amount";
```
It is brittle, and I'm not sure why only that error specifically throws an exception. But keeping it for compatibility.

achow101 commented on src/rpc/rawtransaction_util.h:
```diff
@@ -24,6 +24,7 @@ class SigningProvider;
  * @param result         JSON object where signed transaction results accumulate
  */
 void SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore, const std::map<COutPoint, Coin>& coins, const UniValue& hashType, UniValue& result);
+void MakeSignTransactionResult(CMutableTransaction& mtx, bool complete, const std::map<COutPoint, Coin>& coins, std::map<int, std::string>& input_errors, UniValue& result);
```
Fixed (I think) and renamed

achow101 commented:
> [0eb285f](https://github.com/bitcoin/bitcoin/commit/0eb285f7146ae9beec3a19829cc4420ad9f251ae) `Refactor rawtransaction's SignTransaction into generic SignTransaction function` would be easier to review (`--color-moved=dimmed-zebra -w`) if you renamed the variables in separate commit.

I've undone the renames.

> It may be better to combine [4595326](https://github.com/bitcoin/bitcoin/commit/4595326628bd4c91c753e923535178fca9ade2fd) with [be7991f](https://github.com/bitcoin/bitcoin/commit/be7991f63de2c0cce7d1e622a6ee12dd8f7204c9) so the signing code is moved, rather than just appear (though it's short enough to compare manually).

Squashed them together.

> > Lastly, `GetSigningProvider()` is renamed to `GetPublicSigningProvider()`. It will now only provide pubkeys, key origins, and scripts. `LegacySigningProvider` has it's `GetKey` and `HaveKey` functions changed to only return false. Future implementations should return `HidingSigningProvider`s where private keys are hidden.
> 
> This bit is fairly confusing.

How so? I'm just saying that `GetPublicSigningProvider()` will just return a `SigningProvider` where `GetKey()` and `HaveKey()` will fail.

> This PR is fairly large, and only deduplicates a subset of @vasild's #17577 (only what's needed to move forward with the wallet refactor). I didn't study this other PR in detail, but if it's possible to build on top of that, it would split the review burden and remove more duplication.

I'll take a look at that.

kallewoof reviewed (commented):
Looked at code a bit, but not extensively.

The SignMessage for each scriptpubman is a step backwards considering BIP322 will unify these all into a single place. It would be great if you reviewed that PR (if you already have, and still feel this is the right approach, I'd love to know why).

I'll review/test this PR as well and comment more generally but wanted to put this out there now.

kallewoof commented:
I am trying to merge BIP322 #16440  on top of this, and am running into an issue demonstrated by the below patch:
```diff
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index bde8a6099..dde7819fa 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -559,6 +559,14 @@ static UniValue signmessage(const JSONRPCRequest& request)
     }

     const PKHash *pkhash = boost::get<PKHash>(&dest);
+
+    auto provider = pwallet->GetPublicSigningProvider(GetScriptForDestination(dest));
+    auto keyid = GetKeyForDestination(*provider, dest);
+    CKey secret;
+    if (!provider->GetKey(keyid, secret)) {
+        assert(0);
+    }
+
     if (!pkhash) {
         throw JSONRPCError(RPC_TYPE_ERROR, "Address does not refer to key");
     }
```

The above code hits the assert. I.e. despite getting a signing provider for the destination, it still fails to get the key. This code (i.e. getting signing provider and then getting privkey from that) in current master.

achow101 commented:
> The SignMessage for each scriptpubman is a step backwards considering BIP322 will unify these all into a single place. It would be great if you reviewed that PR (if you already have, and still feel this is the right approach, I'd love to know why).

You can still have the message signing code unified to a single place, and in fact, this PR does that (it consolidates the `signmessage` RPC and the Sign Message dialog signing implementations). Each `ScriptPubKeyMan` can still call that single unified function from within it's own `SignMessage` function. But each one having it's own `SignMessage` also allows us to implement it differently depending on the `ScriptPubKeyMan`. In particular, for hardware wallets, we can't use the unified `SignMessage` function because that would require producing a `SigningProvider` containing private keys, which is not possible for hardware wallets. In order to sign anything with hardware wallets, we need to pass in the thing to be signed to the device and let it handle the signing rather than signing it ourselves.

> The above code hits the assert. I.e. despite getting a signing provider for the destination, it still fails to get the key. This code (i.e. getting signing provider and then getting privkey from that) in current master.

That's because you are still using the old code and way of signing messages. There is no need for `signmessage` to get a `SigningProvider` anymore. Instead you should be implementing either within the `SignMessage` function that is now in `sign.cpp`. You'll probably need to change the arguments, but you shouldn't need to go to the specific call sites like `signmessage` or the Sign Message dialog.

Specifically the reason this hits the assert is that the `SigningProvider` that is returned only returns public keys and scripts. `GetKey` and `HaveKey` have been changed to always fail. Notice that the function named `GetSigningProvider` was renamed to `GetPublicSigningProvider`.

achow101 committed:
Move direct calls to MessageSign into new SignMessage functions in CWallet and ScriptPubKeyMan

Instead of getting a SigningProvider and then going to MessageSign,
have ScriptPubKeyMan handle the message signing internally.

achow101 commented:
I've based this on top of #17577

achow101 commented on src/util/error.h:
```diff
@@ -34,6 +34,14 @@ enum class TransactionError {
 
 std::string TransactionErrorString(const TransactionError error);
 
+enum class SigningError {
```
Renamed to `SigningResult`

kallewoof commented:
> That's because you are still using the old code and way of signing messages. There is no need for signmessage to get a SigningProvider anymore. Instead you should be implementing either within the SignMessage function that is now in sign.cpp. You'll probably need to change the arguments, but you shouldn't need to go to the specific call sites like signmessage or the Sign Message dialog.

Ahh. In that case `PublicSigningProvider` is a bad name for it, I think? It's not providing any *signing* at all, after all.

achow101 commented:
> Ahh. In that case `PublicSigningProvider` is a bad name for it, I think? It's not providing any _signing_ at all, after all. I assume this is used to verify signatures? Maybe `GetPublicKeyProvider`.

It's used primarily for checking solvability. So I've renamed it to `GetSolvingProvider`

Sjors commented:
`GetSolutionProvider`? :-P

instagibbs commented:
concept ACK

Sjors reviewed (commented):
ACK ba91cc878ad18e457c79ca3f84be3b53f1182887 (pending parent PR)

I made a commit that introduces `SolvingProvider` as the `SigningProvider` base class, which I think makes things more clear. But it's lot of code changes: https://github.com/Sjors/bitcoin/commit/7f97d83ccbc291d0a69c7c8aa5b8f9d217d6badf

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2403,34 +2403,66 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+        }
+
+        // Add this spk_man to visited_spk_mans so we can skip it later
+        visited_spk_mans.insert(spk_man->GetID());
```
I considered that, but not all `ScriptPubKeyMan` implementations will be able to sign just one input without potentially signing all inputs. E.g. a hardware wallet spkman. With such spkmans, we would end up signing the entire tx multiple times which is both slow and bad UX.

Sjors replied:
That makes sense. It's worth emphasizing in the comments that `SignTransaction()` will try to sign all inputs, not just `txin`.

achow101 replied:
I've added some comments.

instagibbs commented:
note that this PR is built on https://github.com/bitcoin/bitcoin/pull/17577

achow101 commented on src/script/sign.cpp:
```diff
@@ -465,3 +465,51 @@ bool IsSegWitOutput(const SigningProvider& provider, const CScript& script)
+            }
+        }
+    }
+    return input_errors.empty();;
```
Fixed

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2511,27 +2516,34 @@ TransactionError CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bool& comp
         }
+
+        // Add this spk_man to visited_spk_mans so we can skip it later
+        visited_spk_mans.insert(spk_man->GetID());
```
Nvm, see https://github.com/bitcoin/bitcoin/pull/18115#discussion_r382673176

instagibbs commented on src/rpc/rawtransaction_util.cpp:
```diff
@@ -272,55 +272,27 @@ void SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
+    UniValue vErrors(UniValue::VARR);
+    for (const auto& err_pair : input_errors) {
+        if (err_pair.second == "Missing amount") {
+            // This particular error needs to be an exception for some reason
```
This comment isn't really helpful for future code readers, just reviewers.

CaptJakk replied:
Is the reason known, or is this a "rain dance" and all we know is that it doesn't work if it isn't an Exception?

achow101 replied:
It's preserving previous behavior of the API which would throw an exception for that particular error instead of putting it into the errors array.

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -916,7 +916,8 @@ class CWallet final : public WalletStorage, private interfaces::Chain::Notificat
     bool FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosInOut, std::string& strFailReason, bool lockUnspents, const std::set<int>& setSubtractFeeFromOutputs, CCoinControl);
-    bool SignTransaction(CMutableTransaction& tx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool SignTransaction(CMutableTransaction& tx) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool SignTransaction(CMutableTransaction& tx, const std::map<COutPoint, Coin>& coins, int sighash, std::map<int, std::string>& input_errors) const;
```
Please document when these two functions are used.

achow101 replied:
Added a comment.

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -1150,6 +1151,7 @@ class CWallet final : public WalletStorage, private interfaces::Chain::Notificat
 
     //! Get the ScriptPubKeyMan for a script
     ScriptPubKeyMan* GetScriptPubKeyMan(const CScript& script) const;
+    ScriptPubKeyMan* GetScriptPubKeyMan(const CScript& script, SignatureData& sigdata) const;
```
Please document how this is used.

achow101 replied:
Added a comment.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2403,34 +2403,68 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
-        nIn++;
+
+        // Input needs to be signed, find the right ScriptPubKeyMan
+        ScriptPubKeyMan* spk_man = GetScriptPubKeyMan(coin->second.out.scriptPubKey, sigdata);
```
Will one SPKM always be sufficient for any input even going forward? Seems to me the same SPKM will always be sourced for a particular input?

instagibbs replied:
5cab890b915fd9f25cada7ddab7cc00eea22c80b suggests no, there may be multiple.

achow101 replied:
The expectation is that one spkman will be sufficient. But I suppose it is possible that multiple spkmans could sign for a particular input. I don't think we handle that case.

5cab890 was from a previous iteration where all spkmans signed the transaction. Perhaps it isn't needed anymore.

instagibbs replied:
I think it makes sense to remove that commit then, and just state the assumption here.

achow101 replied:
Dropped the commit. Added a comment about the assumption.

achow101 replied:
I've had to restore the commit. It actually is needed. It actually doesn't imply that the different `ScriptPubKeyMan`s can successfully sign the same input.

It is still needed because each time the wrong `ScriptPubKeyMan` signs an input, it will add an error. Once that input becomes signed by the correct `ScriptPubKeyMan`, we need to clear away those errors otherwise they will incorrectly persist. Thus that commit is still needed.

Sjors replied:
Shouldn't this loop over all matching `GetScriptPubKeyMan`s? Otherwise the comment above should be changed to "We assume that each input is **matched** (and signed) by only one ScriptPubKeyMan.".

Right now:
* if non-matching `ScriptPubKeyMan` tries and fails to sign this input, we'll try with the matching `ScriptPubKeyMan`, but;
* if we match, but can't sign, that error can only be cleared if another input happens to match a `ScriptPubKeyMan` that then happens to be able for the failed input as well.

achow101 replied:
Fine. I've changed it to check all `ScriptPubKeyMan`s that `CanProvide`.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2513,27 +2520,34 @@ TransactionError CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bool& comp
 
-        complete &= SignPSBTInput(HidingSigningProvider(provider.get(), !sign, !bip32derivs), psbtx, i, sighash_type);
-    }
+        // If we've already been signed by this spk_man, skip it
```
Will one SPKM always be sufficient for any input even going forward? Seems to me the same SPKM will always be sourced for a particular input?

achow101 replied:
Documented the assumption.

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -1178,8 +1178,8 @@ class CWallet final : public WalletStorage, private interfaces::Chain::Notificat
     //! Get the SigningProvider for a script
-    std::unique_ptr<SigningProvider> GetSigningProvider(const CScript& script) const;
-    std::unique_ptr<SigningProvider> GetSigningProvider(const CScript& script, SignatureData& sigdata) const;
+    std::unique_ptr<SigningProvider> GetSolvingProvider(const CScript& script) const;
```
A little confused by this commit. Should `SigningProvider` itself be renamed as well?

achow101 replied:
We should probably refactor `SigningProvider` at some point to have a parent `SolvingProvider` that doesn't have private keys. But for now, this just returns a `SigningProvider` without private keys.

Sjors replied:
I have commit for exactly that, see my earlier comment: https://github.com/bitcoin/bitcoin/pull/18115#pullrequestreview-362040407

instagibbs reviewed (approved):
reviewed to 5cab890b915fd9f25cada7ddab7cc00eea22c80b , just a few points, otherwise looks good

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/18115/commits/2525db8eb1af79f00899cac1bd2ff8a1e85d32f8

Sjors commented:
utACK b26777cb7b3fa59c4b518d953165fdb80ca444a9, modulo the below issue

Sjors reviewed (commented):
utACK b26777cb7b3fa59c4b518d953165fdb80ca444a9, modulo the below issue

Sjors commented:
utACK 1f9a806a0184ebc4fe7b7b99df6341e0a00fef80

instagibbs commented on src/script/sign.cpp:
```diff
@@ -509,6 +509,9 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
             }
+        } else {
+            // If this input succeeds, make sure there is no error set for it
+            input_errors.erase(i);
```
Help me understand:

1) Two SPKM: A and B
2) Two input tx:
input 1: key resides in A
input 2: key resides in B
3) A asked to sign 1, successfully
4) A asked to sign 2, unsuccessful, appends error C
5) B asked to sign 1, unsuccessful, appends error D
6) B asked to sign 2, successful, deletes C

We're left with error D?

achow101 replied:
5. B asked to sign 1. 1 comes with existing signature. Skips signing because already signed, no error.

instagibbs replied:
`CWallet::SignTransaction` outer loop over inputs has the complete check, not the inner SPKM loop?

Sjors replied:
@achow101 `CWallet` skips B when considering input 1, because it's already signed. But then when it's considering input 2 it will tell B to sign the transaction, and the sign code doesn't seem skip already signed inputs.

It should be easy to reproduce this scenario in a test on the descriptor wallet PR.

achow101 replied:
To be more specific, the generic `SignTransaction` function uses `DataFromTransaction` on each input to pull out any existing signatures, pubkeys, and scripts into a `SignatureData` object. When this `SignatureData` object gets passed into `ProduceSignature`, `ProduceSignature` first looks at the `SignatureData` finds an existing signatures, pubkeys, and scripts, and fills those into the transaction rather than asking for the `SigningProvider` (which is produced by the `ScriptPubKeyMan`). So even if that `SigningProvider` can't sign, because we already had signatures, pubkeys, and scripts in that transaction, no error occurs.

There is an additional optimization already in `ProduceSignature` which skips signing when it finds that `SignatureData.complete == true`, which is determined by doing a `VerifyScript` within `DataFromTransaction`. Hence signing is skipped for that input.

All of this is pre-existing code which allows us to sign a transaction with external inputs without error. The same process that we do internally with multiple `ScriptPubKeyMan`s is almost the same as if you signed a transaction that has inputs from multiple wallets with those different wallets. Errors are not given for the inputs that are already signed.

instagibbs replied:
Ok, I agree with what you're saying, though it's indirect.

Sjors replied:
That makes sense and will do for now. Would be good to add a test to #15764 for this type of stuff, because it seems brittle.

achow101 replied:
The nature of descriptor wallets as implemented in #16528 requires this commit or tests fail. We are already testing this case. I added back this commit *because* tests were failing without it in #16528.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/18115/commits/1f9a806a0184ebc4fe7b7b99df6341e0a00fef80

Though I wish the code regarding the last commit was easier to grok about.

Empact reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/commit/1f9a806a0184ebc4fe7b7b99df6341e0a00fef80

achow101 commented on src/qt/signverifymessagedialog.cpp:
```diff
@@ -133,20 +133,27 @@ void SignVerifyMessageDialog::on_signMessageButton_SM_clicked()
+    QString error;
+    switch (res) {
+        case SigningResult::OK:
+            error = tr("No error");
```
Meh. Just in case some other change results in `error` being shown.

achow101 commented on src/primitives/transaction.h:
```diff
@@ -124,6 +124,11 @@ class CTxIn
+    {
+        return a.prevout < b.prevout;
+    }
+
```
It looks like this isn't needed at all... But I would like to avoid invalidating ACKs, so I'll leave it in unless I have to rebase or make other changes.

achow101 replied:
Had to rebase, removed

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2926,22 +2926,9 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
-                }
-
-                nIn++;
+            if (!SignTransaction(txNew)) {
```
Will collapse if rebase or other changes are needed.

achow101 replied:
had to rebase, collapsed.

achow101 commented:
There was a hidden merge conflict that required a rebase.

meshcollider reviewed (commented):
Light code review / build / test run ACK 08b9b68528b4644bce6bb925d6b0cbcaae62c5a9

If Empact, instagibbs, or Sjors could post-rebase re-ACK I'll merge this

Sjors commented:
Will do after @achow101 fixes the Travis build failure :-) https://travis-ci.org/bitcoin/bitcoin/jobs/659844515#L2418

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/18115/commits/d2774c09cfcc6c5c967d40bb094eabc8c0bdb6bf

achow101 commented:
> Will do after @achow101 fixes the Travis build failure :-) https://travis-ci.org/bitcoin/bitcoin/jobs/659844515#L2418
> 
> Marking `fillPSBT()` as `const` in `src/interfaces/wallet.{h,cpp}` might help, and would at least be consistent.

I did that, but I don't think it will help. I'm not sure what would cause that travis failure; it's failing on something that the other targets are compiling too. I think travis may have gotten confused?

Sjors commented:
re-utACK d2774c09cfcc6c5c967d40bb094eabc8c0bdb6bf

meshcollider commented:
re-utACK d2774c09cfcc6c5c967d40bb094eabc8c0bdb6bf
